### PR TITLE
core: rpc alloc check memory is valid

### PR DESCRIPTION
**Pull request by jenswi-linaro**
thread_rpc_alloc_arg() and thread_rpc_alloc_payload() only succeeds if
the allocated shared memory is valid shared memory.

Tested-by: Jens Wiklander <jens.wiklander@linaro.org> (QEMU)
Signed-off-by: Jens Wiklander <jens.wiklander@linaro.org>

### CODE REVIEW

File: core/arch/arm/kernel/thread.c [hunk 1]
```diff
@@ -1179,17 +1179,74 @@ uint32_t thread_rpc_cmd(uint32_t cmd, size_t num_params,
...
+	pa = reg_pair_to_64(rpc_args[1], rpc_args[2]);
+	co = reg_pair_to_64(rpc_args[4], rpc_args[5]);
+	if (!check_alloced_shm(pa, size, 8)) {
+		thread_rpc_free_arg(co);
+		pa = 0;
+		co = 0;
+	}
+
 	*arg = reg_pair_to_64(rpc_args[1], rpc_args[2]);
 	*cookie = reg_pair_to_64(rpc_args[4], rpc_args[5]);
```
Comment (etiennec-linaro) [inline core/arch/arm/kernel/thread.c]:
```
*arg = pa;
*cookie = co;
```

File: core/arch/arm/kernel/thread.c [hunk 2]
```diff
@@ -1179,17 +1179,74 @@ uint32_t thread_rpc_cmd(uint32_t cmd, size_t num_params,
...
+	paddr_t pa;
+	uint64_t co;
 	uint32_t rpc_args[THREAD_RPC_NUM_ARGS] = {
 		OPTEE_SMC_RETURN_RPC_ALLOC, size};
 
 	thread_rpc(rpc_args);
+
+	pa = reg_pair_to_64(rpc_args[1], rpc_args[2]);
+	co = reg_pair_to_64(rpc_args[4], rpc_args[5]);
+	if (!check_alloced_shm(pa, size, 8)) {
```
Comment (etiennec-linaro) [inline core/arch/arm/kernel/thread.c]: clarity: could replace `8` with some `sizeof(uint64_t)`

File: core/arch/arm/kernel/thread.c (hunk 1 above)
Comment (jenswi-linaro) [inline core/arch/arm/kernel/thread.c]: Of course, thanks!

File: core/arch/arm/kernel/thread.c (hunk 2 above)
Comment (jenswi-linaro) [inline core/arch/arm/kernel/thread.c]: I'll fix.

**Comment by jenswi-linaro** [comment]
Update

File: core/arch/arm/kernel/thread.c [hunk 3]
```diff
@@ -1179,14 +1179,71 @@ uint32_t thread_rpc_cmd(uint32_t cmd, size_t num_params,
...
+	if (pa & (align - 1))
+		return false;
+	return core_pbuf_is(CORE_MEM_NSEC_SHM, pa, len);
+}
+
+void thread_rpc_free_arg(uint64_t cookie)
+{
+	if (cookie) {
+		uint32_t rpc_args[THREAD_RPC_NUM_ARGS] = {
+			OPTEE_SMC_RETURN_RPC_FREE};
```
Comment (etiennec-linaro) [inline core/arch/arm/kernel/thread.c]: minor: missing before at `}`. check checkpatch.

**Comment by etiennec-linaro** [comment]
`Reviewed-by: etienne carriere <etienne.carriere@linaro.org>`

File: core/arch/arm/kernel/thread.c (hunk 3 above)
Comment (jenswi-linaro) [inline core/arch/arm/kernel/thread.c]: I'll fix.

**Comment by jenswi-linaro** [comment]
Tag applied

**Comment by jforissier** [comment]
`checkpatch` has one warning.

**Comment by jenswi-linaro** [comment]
Fixed checkpatch warning

**Pull request merged**